packages/storage-plus/src/de.rs

```diff
@@ -0,0 +1,157 @@
+use crate::keys::IntKey;
+use cosmwasm_std::{Addr, StdError, StdResult};
+use std::array::TryFromSliceError;
+use std::convert::TryInto;
+
+pub trait Deserializable {
+    type Output: Sized;
+
+    fn from_slice(value: &[u8]) -> StdResult<Self::Output>;
+    fn from_vec(value: &Vec<u8>) -> StdResult<Self::Output>;
+}
+
+impl Deserializable for () {
+    type Output = ();
+
+    fn from_slice(_value: &[u8]) -> StdResult<Self::Output> {
+        Ok(())
+    }
+    fn from_vec(_value: &Vec<u8>) -> StdResult<Self::Output> {
+        Ok(())
+    }
+}
+
+macro_rules! bytes_de {
+    (for $($t:ty),+) => {
+        $(impl Deserializable for $t {
+            type Output = Vec<u8>;
+
+            fn from_slice(value: &[u8]) -> StdResult<Self::Output> {
+                Ok(value.to_vec())
+            }
+            fn from_vec(value: &Vec<u8>) -> StdResult<Self::Output> {
+                Ok(value.to_vec())
+            }
+        })*
+    }
+}
+
+bytes_de!(for Vec<u8>, &Vec<u8>, [u8], &[u8]);
+
+macro_rules! string_de {
+    (for $($t:ty),+) => {
+        $(impl Deserializable for $t {
+            type Output = String;
+
+            fn from_slice(value: &[u8]) -> StdResult<Self::Output> {
+                // FIXME?: Use `from_utf8_unchecked` for String, &str
+                String::from_utf8(value.to_vec())
+                    // FIXME: Add and use StdError utf-8 error From helper
+                    .map_err(|err| StdError::generic_err(err.to_string()))
+            }
+
+            fn from_vec(value: &Vec<u8>) -> StdResult<Self::Output> {
+                // FIXME?: Use `from_utf8_unchecked` for String, &str
+                String::from_utf8(value.to_vec())
+                    // FIXME: Add and use StdError utf-8 error From helper
+                    .map_err(|err| StdError::generic_err(err.to_string()))
+            }
+        })*
+    }
+}
+
+string_de!(for String, &String, str, &str, Addr, &Addr);
+
+macro_rules! integer_de {
+    (for $($t:ty),+) => {
+        $(impl Deserializable for IntKey<$t> {
+            type Output = $t;
+
+            fn from_slice(value: &[u8]) -> StdResult<Self::Output> {
+                Ok(<$t>::from_be_bytes(value.try_into()
+                    // FIXME: Add and use StdError try-from error From helper
+                    .map_err(|err: TryFromSliceError| StdError::generic_err(err.to_string()))?))
+            }
+
+            fn from_vec(value: &Vec<u8>) -> StdResult<Self::Output> {
+                Ok(<$t>::from_be_bytes(value.to_vec().try_into().unwrap()))
```

Comment:
And why does this unwrap rather than map_err like above?

Reply:
Good call, I forgot about it. This is just because of wip.